Typescript improvements

williamlardier opened the pull request:
Improve TS:

- Auth types are fixed as they were inconsistent. We still have a mix of 4 possible types as output whtn doing auth (authinfo, authv4 info, authv2 or public user).
- Old style `_` prefix to define private variable in TS classes was removed: this is not following the TS standard. More file might still be affected.
- Data models for both Bucket and Objects were standardized to avoid using `any`.
  - This allowed to find a limitation in the createBucket function in the mongo client interface.

Issue: ARSN-462

Seragonia committed:
Improve models types

Issue: ARSN-462

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-462 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.2.4`


Please check the `Fix Version/s` of ARSN-462, or the target
branch of this pull request.

Seragonia committed:
Improve AuthV2 and AuthV4 types

Issue: ARSN-462

francoisferrand commented on lib/auth/auth.ts:
```diff
@@ -145,16 +145,12 @@ function doAuth(
-    if (res.params.version === 2) {
-        // @ts-ignore
-        return vault!.authenticateV2Request(res.params, requestContexts, cb);
+    res.params!.log = log;
```
do you know why Typescript is not able to infer the type here?
As far as I understand, `extractParams()` return either AuthV2RequestParams, AuthV4RequestParams or AuthInfo. Since the AuthInfo case is handle, TS should be able to deduce this will be either AuthV2RequestParams or AuthV4RequestParams...

Maybe this is due to the "optional" params (`params?`) in result of extractParams?
And the return type should be better type to either return an error without params, or params with null error: so that type deduction can work its magic, and we don't need type coercion...

williamlardier replied:
I updated the PR, I created a new type that allows TS to understand that params is defined if `err` is `null`.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-462 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.2.5`


Please check the `Fix Version/s` of ARSN-462, or the target
branch of this pull request.

francoisferrand commented on lib/auth/auth.ts:
```diff
@@ -61,7 +61,7 @@ function extractParams(
     awsService: string,
     data: { [key: string]: string }
-) {
+): { err: null | Error | ArsenalError, params?: AuthV2RequestParams | AuthV4RequestParams | AuthInfo } {
```
can we really return an `Error` ?

williamlardier replied:
Through the `checkFunctions`, yes: the `validateCredentials` function in the header+query v4 check can return such error. V2 isn't returning `Error`s indeed.

francoisferrand replied:
`validateCredentials()` function is typed to return an `Error`, but implementation looks like it is always returning an `ArsenalError`

williamlardier replied:
I updated the code (and all other related comments).

francoisferrand commented on lib/auth/v2/headerAuthCheck.ts:
```diff
@@ -1,11 +1,16 @@
+    request: any,
+    log: Logger,
+    data: { [key: string]: string },
+): { err: null | ArsenalError | Error, params?: AuthV2RequestParams } {
```
can we really return an `Error`, are not just missing some typing below?

francoisferrand commented on lib/auth/v2/queryAuthCheck.ts:
```diff
@@ -1,14 +1,19 @@
+    request: any,
+    log: Logger,
+    data: { [key: string]: string },
+): { err: null | ArsenalError | Error, params?: AuthV2RequestParams } {
```
can we really return an Error, are not just missing some typing below?

francoisferrand commented on lib/auth/v4/constructStringToSign.ts:
```diff
@@ -40,14 +40,6 @@ export default function constructStringToSign(params: {
         service: params.awsService,
     });
 
-    // TODO Why that line?
```
why indeed?
and why can we now remove it?

williamlardier replied:
I removed it because after reviewing the function and checking all the types, the function cannot return an error, so this code cannot be reached.

francoisferrand commented on lib/auth/v4/headerAuthCheck.ts:
```diff
@@ -26,7 +27,7 @@ export function check(
     data: { [key: string]: string },
     awsService: string
-) {
+): { err: null | ArsenalError | Error, params?: AuthV4RequestParams } {
```
It seems this cannot return an `Error`.

The case would be a failure to utf8.encode:
```
    const stringToSign = constructStringToSign({ ... });
    if (stringToSign instanceof Error) {
        return { err: stringToSign };
    }
```

But according to the docs this function does not return an error, it throws on error:

> Encodes any given JavaScript string (string) as UTF-8, and returns the UTF-8-encoded version of the string. It throws an error if the input string contains a non-scalar value, i.e. a lone surrogate. (If you need to be able to encode non-scalar values as well, use WTF-8 instead.)

francoisferrand commented on lib/auth/v4/queryAuthCheck.ts:
```diff
@@ -1,18 +1,23 @@
+    request: any,
+    log: Logger,
+    data: { [key: string]: string },
+): { err: null | ArsenalError | Error, params?: AuthV4RequestParams } {
```
same

francoisferrand commented on lib/models/BucketPolicy.ts:
```diff
@@ -48,64 +48,77 @@ const objectActions = [
-    _json: string;
-    _policy: any;
+    private json: string;
+    private policy: BucketPolicyMetadata | null = null;
```
would this not be more appropriate (and consistent with the changes above)?
```suggestion
    private policy?: BucketPolicyMetadata;
```

francoisferrand commented on lib/models/WebsiteConfiguration.ts:
```diff
@@ -77,10 +77,10 @@ export type RedirectAllRequestsTo = {
+    indexDocument?: string;
+    errorDocument?: string;
+    redirectAllRequestsTo?: RedirectAllRequestsTo;
+    routingRules?: RoutingRule[] | any[],
```
does not seem really typed....

williamlardier replied:
I updated it, due to a wrong type being used in BucketInfo (class type instead of the "Params" version)

williamlardier replied:
Actually, the whole website implementation is a mess: js code uses private members of the website class and we eneded up storing that in mongodb, so we need to keep this behavior... Working on it

williamlardier replied:
Done...

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -407,7 +407,11 @@ class MongoClientInterface {
-                value: newBucketMD,
+                value: {
+                    ...newBucketMD,
+                    quotaMax: new Long(newBucketMD.quotaMax || '0'),
```
should have an associated test, to ensure we can indeed properly create buckets with very large quotas

francoisferrand replied:
> it fails the type validation

what type validation? payload is untyped (for typescript), right?

> Do we still want such test?

from MongoClientInterface standpoint, it is possible to set a quota (i.e. it uses the `newBucketMD.quotaMax`) which is provided: so i think it would be better indeed to add one (or more: no quota/small quota/large bigint) test case in MongoClientInterface tests

williamlardier replied:
> what type validation? payload is untyped (for typescript), right?

We enforce types in the mongodb driver, o we cannot actually create a bucket MD if the quota is not a Long (which is expected). I'll add a test

Seragonia committed:
Improve typing post review

Issue: ARSN-462

Seragonia committed:
Improve logger types after rebase

Issue: ARSN-462

ghost commented:
I added a commit to fix the types of the loggers and fix a conflict with Vault.ts.

BourgoisMickael commented on lib/auth/v2/headerAuthCheck.ts:
```diff
@@ -1,11 +1,17 @@
+export function check(
+    request: any,
+    log: RequestLogger,
+    data: { [key: string]: string },
```
Don't you prefer:

```suggestion
    data: Record<string, string>,
```

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-462 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.2.6`


Please check the `Fix Version/s` of ARSN-462, or the target
branch of this pull request.

BourgoisMickael commented on lib/auth/auth.ts:
```diff
@@ -1,6 +1,6 @@
 import * as crypto from 'crypto';
-import { Logger } from 'werelogs';
-import errors from '../errors';
+import { RequestLogger } from 'werelogs';
```
```suggestion
import type { RequestLogger } from 'werelogs';
```

For type only import this can be interesting as the import is discarded after transpiling

BourgoisMickael commented on lib/models/BucketAzureInfo.ts:
```diff
@@ -3,32 +3,34 @@ export type DeleteRetentionPolicy = {
-        minuteMetrics: any;
-        serviceVersion: string;
-    };
+    private data: AzureInfoMetadata;
```
```suggestion
    private readonly data: AzureInfoMetadata;
```

And this type can be used in the constructor as well ?

BourgoisMickael commented on lib/models/BucketPolicy.ts:
```diff
@@ -48,64 +48,77 @@ const objectActions = [
+    Statement: {
+        Sid: string;
+        Effect: string;
+        Principal: string | { AWS: string[] };
```
```suggestion
        Principal: string | { AWS: string | string[] };
```

This can be a string as well like `Principal: { AWS: '*' }`

BourgoisMickael commented on lib/models/BucketPolicy.ts:
```diff
@@ -48,64 +48,77 @@ const objectActions = [
+    Id: string;
+    Version: string;
+    Statement: {
+        Sid: string;
```
`Id` and `Sid` should be optional

BourgoisMickael commented on lib/models/BucketPolicy.ts:
```diff
@@ -48,64 +48,77 @@ const objectActions = [
+    Version: string;
+    Statement: {
+        Sid: string;
+        Effect: string;
```
```suggestion
        Effect: 'Allow' | 'Deny';
```

BourgoisMickael commented on lib/models/BucketPolicy.ts:
```diff
@@ -48,64 +48,77 @@ const objectActions = [
+        Principal: string | { AWS: string[] };
+        Action: string | string[];
+        Resource: string | string[];
+        Condition?: { [key: string]: string | string[] };
```
There is a missing intermediate object and value can be number
```suggestion
        Condition?: Record<string, Record<string, string | string[] | number>>;
```

Example 

```json

{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Deny",
            "Principal": "*",
            "Action": "s3:PutObject",
            "Resource": [
                "arn:aws:s3:::amzn-s3-demo-bucket1",
                "arn:aws:s3:::amzn-s3-demo-bucket1/*"
            ],
            "Condition": {
                "NumericLessThan": {
                    "s3:TlsVersion": 1.2
                }
            }
        }
    ]
}
```

BourgoisMickael commented on lib/models/WebsiteConfiguration.ts:
```diff
@@ -22,7 +22,7 @@ export type Condition = {
 };
 
-export type RoutingRuleParams = { redirect: Redirect; condition?: Condition };
+export type RoutingRuleParams = { redirect?: Redirect; condition?: Condition };
```
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/aws-sdk2-types/clients/s3.d.ts#L6279

I don't see the redirect being optional

Seragonia committed:
Improve auth type post addition of options

Issue: ARSN-462

Seragonia committed:
Apply typing suggestions from review

- Prefer using Record type
- Use Record everywhere when applicable
- Improve object lock logic and types

Issue: ARSN-462

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-462 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.2.7`


Please check the `Fix Version/s` of ARSN-462, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

ghost commented:
Note for reviewers: all comments adressed & the PR was updated after #2288 merge, adding more types for request contexts and auth options.

BourgoisMickael commented on lib/policyEvaluator/RequestContext.ts:
```diff
@@ -154,7 +154,31 @@ export type RequesterInfo = {
+};
+
+export type RequestContextType = {
+    headers: { [key: string]: string | string[] };
```
```suggestion
    headers: Record<string, string | string[]>;
```

williamlardier replied:
thanks! I added https://github.com/scality/Arsenal/pull/2299/commits/5542f1ed0649c2509aa6c0756445a3923003064c to do it everywhere in Arsenal.

francoisferrand commented on lib/models/ObjectLockConfiguration.ts:
```diff
@@ -89,23 +100,18 @@ export default class ObjectLockConfiguration {
             return { error };
         }
-        const timeType = 'Days' in dr ? 'Days' : 'Years';
+        const timeType = dr.Days ? 'Days' : 'Years';
         if (!dr[timeType] || !dr[timeType][0]) {
             const msg = 'request xml does not contain Days or Years';
```
looking at this code, we should push the cleanup a bit further I think: the variable we actually need is `dr[timeType]`, and we can avoid accessing field by name (and later type assertion)

```
if (dr.Days && dr.Years) {
    return 'request xml contains both Days and Years';
}
if (!dr.Days && !dr.Years) {
    return 'request xml does not contain Days or Years';
}
const {timeRaw, timeMax, timeType} = func() {
    if (dr.Days) {
        return {dr.Days, 36500, 'days'};
    } else {
        return {dr.Years, 100, 'years'};
    }
}(); 
const timeValue = Number.parseInt(timeRaw[0], 10);
```

francoisferrand commented on lib/models/ObjectLockConfiguration.ts:
```diff
@@ -89,23 +100,18 @@ export default class ObjectLockConfiguration {
             const error = errors.MalformedXML.customizeDescription(msg);
             return { error };
         }
-        if (dr[timeType].length > 1) {
```
why remove this check?

it is consistent with the array indexing we do (`dr[timeType][0]`), though I don't understand this indexing with regard to the typing: it seems to me either the type is not correct (we can an array of string, not a string), or the indexing is just taking the first digit?

williamlardier replied:
So it's indeed a string array, and it's due to how the xml parser works. The old code is correct, but the type wasn't. I'll update and add your suggestions.

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -36,7 +37,7 @@ export type SSE = {
 
 export type VersioningConfiguration = {
     Status: string;
```
should be typed as well (can be Enabled, Disabled or Suspended I think) ?

francoisferrand commented on lib/auth/auth.ts:
```diff
@@ -1,6 +1,6 @@
-import { Logger } from 'werelogs';
-import errors from '../errors';
+import type { RequestLogger } from 'werelogs';
+import errors, { ArsenalError } from '../errors';
```
only need type of `ArsenalError` ?

francoisferrand commented on lib/auth/auth.ts:
```diff
@@ -118,39 +131,38 @@ function extractParams(
+            if (data.authType) {
                 requestContext.setAuthType(data.authType);
+            }
+            if (data.signatureVersion) {
                 requestContext.setSignatureVersion(data.signatureVersion);
+            }
+            if (data.securityToken) {
                 requestContext.setSecurityToken(data.securityToken);
-                if ('signatureAge' in data) {
-                    requestContext.setSignatureAge(data.signatureAge);
-                }
             }
+            requestContext.setSignatureAge(data.signatureAge || 0);
```
- this will crash if `res.params.data` is not set, which was not the case before
- we used not to call `setSignatureAge` when `data.signatureAge` was not set

```suggestion
            if (data?.authType) {
                requestContext.setAuthType(data.authType);
            }
            if (data?.signatureVersion) {
                requestContext.setSignatureVersion(data.signatureVersion);
            }
            if (data?.securityToken) {
                requestContext.setSecurityToken(data.securityToken);
            }
            if (data?.signatureAge) {
                requestContext.setSignatureAge(data.signatureAge);
            }
```
or
```suggestion
    const data = res.params.data;
    if (data && requestContexts) {
        requestContexts.forEach(requestContext => {
            if (data.authType) {
                requestContext.setAuthType(data.authType);
            }
            if (data.signatureVersion) {
                requestContext.setSignatureVersion(data.signatureVersion);
            }
            if (data.securityToken) {
                requestContext.setSecurityToken(data.securityToken);
            }
            if (data.signatureAge) {
                requestContext.setSignatureAge(data.signatureAge);
            }
        }
    }
```

francoisferrand commented on lib/auth/Vault.ts:
```diff
@@ -63,24 +63,38 @@ export function vaultSignatureCb(
+    log: RequestLogger;
+    data: {
+        accessKey: string;
+        algo: string;
```
can we type `algo`, with the possible values?

francoisferrand commented on lib/auth/Vault.ts:
```diff
@@ -63,24 +63,38 @@ export function vaultSignatureCb(
     data: {
         accessKey: string;
-        signatureFromRequest: string;
+        algo?: string;
```
can we type algo, with the possible values?

francoisferrand commented on lib/auth/Vault.ts:
```diff
@@ -63,24 +63,38 @@ export function vaultSignatureCb(
-        signatureFromRequest: string;
+        algo?: string;
+        authType?: 'query' | 'header' | 'REST-HEADER' | 'REST-QUERY-STRING';
+        credentialScope?: string;
```
can we type `credentialScope`, with the possible values?

williamlardier replied:
We cannot, this is dynamically extracted from the string to sign.

francoisferrand commented on lib/auth/v4/validateInputs.ts:
```diff
@@ -1,5 +1,5 @@
-import { Logger } from 'werelogs';
-import errors from '../../../lib/errors';
+import type { RequestLogger } from 'werelogs';
+import errors, { ArsenalError } from '../../../lib/errors';
```
only type needed for ArsenalError ?

francoisferrand commented on lib/models/ObjectMDAzureInfo.ts:
```diff
@@ -1,22 +1,24 @@
+export type ObjectMDAzureInfoMetadata = {
```
can we type these fields (containerPublicAccess, containerStoredAccessPolicies, containerImmutabilityPolicy, blobType, blobCopyInfo) ?
should we? (not sure they are still used, it may be a remnant of BlobServer...)

williamlardier replied:
I think we can but as you said this is not used today in cloudserver. I also see several fields with values I cannot easily type (no information, no usage beyond this class). Thinking from a higher level, I think a debt we have here is to potentially remove useless fields from metadata objects (and potentially don't write fully the fields like `'x-amz-server-side-encryption-customer-algorithm'` but use a dict so we greatly reduce the size of MD, in turn reducing storage, RAM consumption, network usage, impact of oplog listening, etc).

I created OS-932.

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/ARSN-462` and the
destination branch `development/8.2`.

Please resolve the conflict on **the feature branch** (`improvement/ARSN-462`).

```sh
git fetch && \
git checkout origin/improvement/ARSN-462 && \
git merge origin/development/8.2
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:improvement/ARSN-462
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

ghost commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.2`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.1`



Please check the status of the associated issue ARSN-462.

Goodbye williamlardier.




*The following options are set:* **approve**
